### DISCUSSION TITLE
Multiple attribute statements

### DISCUSSION THREAD

**Discussion by bigbug** [ts-graphviz/ts-graphviz#796] [Q&A] [answered]
## Describe the bug

Multiple attribute statements regarding the same element (node, edge or graph) are not recognised, stored or outputted by the module.

## To Reproduce

Code to reproduce the behaviour:

Input dot Graph:
```dot
digraph G {
  node [style = filled; fillcolor = "yellow"] a0;
  node [fillcolor = "green"] a1;
  a0 -> a1;
}
```

Code:
```typescript
import { toDot, fromDot } from "ts-graphviz";
const output = toDot(fromDot(dot));
```

Output dot (the yellow color does not appear anymore):
```dot
digraph "G" {
  node [
    style = "filled";
    fillcolor = "green";
  ];
  "a0";
  "a1";
  "a0" -> "a1";
}
```

## Expected behavior

The expected behaviour is to retain the original node colouring. The dot language also explicitly allows this as described [here](https://graphviz.org/doc/info/lang.html).

**Comment by kamiazya** [comment, marked as answer]
Hi @bigbug,

### Behavior Description and Simple Solutions

Yes, this is the behavior seen when converting DOT language ASTs to Model within the `fromDot` function.

This behavior occurs because Model stores attributes in a hash map format, rather than in a data structure that maintains them in the order in which they are listed.

See https://github.com/ts-graphviz/ts-graphviz/blob/main/src/core/models/GraphBase.ts#L39-L43

To avoid this behavior, you must use the `parse` and `stringify` functions provided in the ast module.

```ts
import { parse, stringify } from "ts-graphviz/ast";
const output = stringify(parse(dot));
```

See [Runkit Example](https://runkit.com/embed/ft0gdc4qhjbb).

![img](https://raw.githubusercontent.com/ts-graphviz/ts-graphviz/main/media/state-machine.svg)

### Detailed consultation on expected behavior

I explained above that this is the current expected behavior and how to work around it using the AST module, but I suspect that your use case is more complex than a simple combination of fromDot and toDot functions as in the example.

Am I correct in assuming that you are expecting the case of manipulating the model in the diagram above?

If so, you will need to consider new functions to store the order in which you manipulate the Model and to manipulate the order of precedence during rendering.

Let us know if there is a way we can help you with your use case.


Thanks.

**Comment by bigbug** [comment]
Hello kamiazya, thanks for the fast and very detailed answer. It's very rare to get it this timely and as far me concerns: Props to you!
You are right - it does not happen on parse/stringify. I switched to them and the problem is solved. Sorry for not reading the documentation properly.
I will close the ticket now.

**Comment by kamiazya** [comment]
@bigbug I'm glad I seem to have solved my problem :+1: 

I'll migrate the questions to the GitHub Discussion, as it seems to have a better perspective from others.

**Comment by kamiazya** [comment]
@all-contributors please add @bigbug  for question.

[a comment by an automated account is not shown]
